catalog/spatial/registry/registry-federation-admin-service-impl/src/main/java/org/codice/ddf/registry/federationadmin/service/impl/RegistryPublicationServiceImpl.java

```diff
@@ -0,0 +1,203 @@
+    public void publish(String registryId, String destinationRegistryId)
+            throws FederationAdminException {
+        Metacard metacard = getMetacard(registryId);
+        String metacardId = metacard.getId();
+
+        List<Serializable> locations = new ArrayList<>();
+        Attribute locAttr = metacard.getAttribute(RegistryObjectMetacardType.PUBLISHED_LOCATIONS);
+        if (locAttr != null) {
+            locations = new HashSet<>(locAttr.getValues()).stream()
+                    .map(Object::toString)
+                    .collect(Collectors.toCollection(ArrayList::new));
+        }
+
+        if (locations.contains(destinationRegistryId)) {
+            return;
+        }
+
+        String sourceId = getSourceIdFromRegistryId(destinationRegistryId).orElse(null);
+
+        if (StringUtils.isNotBlank(sourceId)) {
+            federationAdminService.addRegistryEntry(metacard, Collections.singleton(sourceId));
+
+            //need to reset the id since the framework reset it in the groomer plugin
+            //and we don't want to update with the wrong id
+            metacard.setAttribute(new AttributeImpl(Metacard.ID, metacardId));
+
+            locations.add(destinationRegistryId);
+            if (locAttr != null) {
+                locAttr.getValues()
+                        .add(destinationRegistryId);
+            } else {
+                locAttr = new AttributeImpl(RegistryObjectMetacardType.PUBLISHED_LOCATIONS,
+                        locations);
+            }
+            metacard.setAttribute(locAttr);
+            metacard.setAttribute(new AttributeImpl(RegistryObjectMetacardType.LAST_PUBLISHED,
+                    Date.from(ZonedDateTime.now()
+                            .toInstant())));
+
+            federationAdminService.updateRegistryEntry(metacard);
+        }
+
+    }
+
+    @Override
+    public void unpublish(String registryId, String destinationRegistryId)
+            throws FederationAdminException {
+        Metacard metacard = getMetacard(registryId);
+
+        Attribute locAttr = metacard.getAttribute(RegistryObjectMetacardType.PUBLISHED_LOCATIONS);
+        if (locAttr == null || !locAttr.getValues()
+                .contains(destinationRegistryId)) {
+            return;
+        }
+
+        String sourceId = getSourceIdFromRegistryId(destinationRegistryId).orElse(null);
+
+        if (StringUtils.isNotBlank(sourceId)) {
+            federationAdminService.deleteRegistryEntriesByRegistryIds(Collections.singletonList(
+                    registryId), Collections.singleton(sourceId));
+
+            locAttr.getValues()
+                    .remove(destinationRegistryId);
+            metacard.setAttribute(locAttr);
+            metacard.setAttribute(new AttributeImpl(RegistryObjectMetacardType.LAST_PUBLISHED,
+                    Date.from(ZonedDateTime.now()
+                            .toInstant())));
+
+            federationAdminService.updateRegistryEntry(metacard);
+        }
```

Comment:
Changed the logic to throw an exception from getSourceIdFromRegistryId if it doesn't find a source id for the given registryId.